sw/host/opentitanlib/src/util/unknown.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Creates C-like enums which preserve unknown (un-enumerated) values.
+///
+/// If you wanted an enum like:
+/// ```
+/// #[repr(u32)]
+/// pub enum HardenedBool {
+///     True = 0x739,
+///     False = 0x146,
+///     Unknown(u32),
+/// }
+/// ```
+///
+/// Where the `Unknown` discriminator would be the catch-all for any
+/// non-enumerated values, you can use `with_unknown!` as follows:
+///
+/// ```
+/// with_unknown! {
+///     pub enum HardenedBool: u32 {
+///         True = 0x739,
+///         False = 0x14d,
+///     }
+/// }
+/// ```
+///
+/// This "enum" can be used later in match statements:
+/// ```
+/// match foo {
+///     HardenedBool::True => do_the_thing(),
+///     HardenedBool::False => do_the_opposite_thing(),
+///     HardenedBool(x) => panic!("Oh noes! {} is neither True nor False!", x),
+/// }
+/// ```
+///
+/// Behind the scenes, `with_unknown!` implements a newtype struct and
+/// creates associated constants for each of the enumerated values.
+/// The struct also implements `Copy`, `Clone`, `PartialEq`, `Eq`,
+/// `PartialOrd`, `Ord`, `Hash`, `Debug` and `Display` (including the hex,
+/// octal and binary versions).
+///
+/// In addition, `serde::Serialize` and `serde::Deserialize` are
+/// implemented.  The serialized form is a string for known values and an
+/// integer for all unknown values.
+#[macro_export]
+macro_rules! with_unknown {
+    () => {};
```

Comment:
Nitpick: This pattern is now redundant, as the main pattern will match empty arguments and generate zero output.